### PR TITLE
ARROW-17117: [C++] Add timestamp column datatype support for AsOfJoin

### PR DESCRIPTION
**Pull request by iChauster**
Previously, AsOfJoin only supported `int64` as it's `on` column. Here, we add support for `arrow::timestamp(TimeUnit::NANO, "UTC")`.

If there is a more generalized way of dealing with particular timestamps, please let me know!

### CODE REVIEW

File: cpp/src/arrow/compute/exec/asof_join_node.cc [hunk 1]
```diff
@@ -460,10 +464,9 @@ class CompositeReferenceTable {
   }
 
   template <class Builder, class PrimitiveType>
-  Result<std::shared_ptr<Array>> MaterializePrimitiveColumn(MemoryPool* memory_pool,
+  Result<std::shared_ptr<Array>> MaterializePrimitiveColumn(Builder& builder,
```
Comment (icexelloss) [inline cpp/src/arrow/compute/exec/asof_join_node.cc:467]: What is this change for?

File: cpp/src/arrow/compute/exec/asof_join_node.cc [hunk 2]
```diff
@@ -630,24 +650,28 @@ class AsofJoinNode : public ExecNode {
       for (int i = 0; i < input_schema->num_fields(); ++i) {
         const auto field = input_schema->field(i);
         if (field->name() == on_field_name) {
-          if (kSupportedOnTypes_.find(field->type()) == kSupportedOnTypes_.end()) {
-            return Status::Invalid("Unsupported type for on key: ", field->name());
+          // Equals must be used when checking for timestamp types.
+          if (!field->type()->Equals(TimestampType(TimeUnit::NANO, "UTC")) &&
```
Comment (icexelloss) [inline cpp/src/arrow/compute/exec/asof_join_node.cc:654]: Can you put TimestampType in kSupportedOnTypes_

File: cpp/src/arrow/compute/exec/asof_join_node.cc [hunk 3]
```diff
@@ -478,6 +481,23 @@ class CompositeReferenceTable {
...
+                                                            col_index_t i_col) {
+    Builder builder(memory_pool);
+    return MaterializePrimitiveColumn<Builder, PrimitiveType>(builder, i_table, i_col);
+  }
+
+  template <class PrimitiveType>
+  Result<std::shared_ptr<Array>> MaterializeTimestampPrimitiveColumn(
+      MemoryPool* memory_pool, size_t i_table, col_index_t i_col) {
+    TimestampBuilder builder =
+        TimestampBuilder(timestamp(TimeUnit::NANO, "UTC"), memory_pool);
```
Comment (icexelloss) [inline cpp/src/arrow/compute/exec/asof_join_node.cc:497]: Can you not hardcode the timezone here

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 2 above)
Comment (iChauster) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:654]: e.g, even though I have added `TimestampType` to `kSupportedOnTypes_`, it has no effect. I looked for a more generic way to detect a `TimestampType`, like `arrow::istimestamp_type` in `src/arrow/type_traits.h` but I'm not sure how it works with the templating.

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 1 above)
Comment (iChauster) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:467]: Was fighting with the compiler for a bit on this: 
`TimestampBuilder` does not seem to be able to initialized with just a `memory_pool`, I think it requires a `arrow::DataType` object as well. That is the only difference that requires this function to be refactored and split into `MaterializePrimitveColumn` and `MaterializeTimestampPrimitiveColumn`.

**Comment by iChauster** [comment]
@icexelloss , yeah I was hoping it would be easier. I think we would really only need a way to tell if a data type is a timestamp at runtime, and we should be able to supply this information to `TimestampBuilder`.

Worst case scenario, we can enumerate over all possible timestamps (one for each time unit and timezone we'd like to support), but I don't think this is a good solution.

**Comment by iChauster** [comment]
Do we want to do the enumeration, or is supporting this one timestamp (Nanosecond, UTC) okay for now?

**Comment by iChauster** [comment]
@icexelloss I adjusted DoRunBasicTest to run with the same data, but with the timestamp as the time column data type in the schema. Is there anything additional that might interesting to test?

**Comment by icexelloss** [comment]
Let's close this PR. Will pick up and do a new one.

**Comment by rok** [comment]
> Do we want to do the enumeration, or is supporting this one timestamp (Nanosecond, UTC) okay for now?

It would probably be best to support temporal, integer and perhaps(?) float types.

> Let's close this PR since this is the last week of Ivan's internship. Will pick this up and open a new one.

Nice work @iChauster!
Feel free to ping me on the new PR @icexelloss.

**Pull request closed without merging**